servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/MpmcSequentialRunQueue.java

```diff
@@ -0,0 +1,185 @@
+    private static final AtomicReferenceFieldUpdater<MpmcSequentialRunQueue, Node>
+            tailUpdater = newUpdater(MpmcSequentialRunQueue.class, Node.class, "tail");
+
+    @Nullable
+    private volatile Node tail;
+
+    /**
+     * Offer {@link Node} to this queue.
+     * @param node The {@link Node} to append. Must only exclusively be offered to this queue instance.
+     */
+    void offer(final Node node) {
+        for (;;) {
+            final Node currentTail = this.tail;
+            if (currentTail == null) {
+                if (tailUpdater.compareAndSet(this, null, node)) {
+                    // node has been inserted and is the only node, we initiate processing.
+                    node.run();
+                    break;
+                }
+                // Another thread won the race to offer a node, loop around and try again.
+            } else if (currentTail.append(node)) {
+                // Make the newly appended node visible as the tail. This is a best effort CAS and may fail because:
+                // 1. Another thread is also inserting, has a stale tail, followed its existing tail links, and updated
+                // the tail reference via offerPatchTail.
+                // 2. The consumer thread has seen the link from the old tail to the new node, processed node,
+                // popped node from the list (updated node's next to point to EMPTY_NODE), another producer thread
+                // appends a new node, sees the tail is popped, and updates the tail reference via CAS.
+                tailUpdater.compareAndSet(this, currentTail, node);
+                break;
+            } else if (currentTail.isPopped()) {
+                // A previously appended node was processed, and popped before updating the tail after append. In that
+                // case the tail maybe pointing to an invalid node and we clear it out.
+                if (tailUpdater.compareAndSet(this, currentTail, node)) {
+                    node.run();
+                    break;
+                }
+                // Best effort to clear the tail, and failure is OK because:
+                // 1. Another thread is in offer and already patched up the tail pointer and we will read the new tail
+                // on the next loop iteration.
+            } else {
+                // We failed to append to currentTail's next pointer, which means currentTail isn't the tail, and
+                // currentTail hasn't been popped. This means the tail pointer is stale, so we re-read the reference
+                // to see if it changed, and if not attempt to walk the list and update the tail pointer.
+                final Node newTail = this.tail;
+                if (newTail == currentTail && offerUpdateStaleTail(node, currentTail)) {
+                    break;
+                }
+            }
+        }
+    }
+
+    private boolean offerUpdateStaleTail(final Node node, final Node oldTail) {
```

Comment:
same as above... I think the method name is a bit odd as it may queue but also may run.